src/Features/Core/Portable/BracePairs/IBracePairsService.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.BracePairs
+{
+    internal readonly record struct BracePairData(
+        TextSpan Start,
+        TextSpan End);
+
+    internal interface IBracePairsService : ILanguageService
+    {
+        Task AddBracePairsAsync(Document document, TextSpan textSpan, ArrayBuilder<BracePairData> bracePairs, CancellationToken cancellationToken);
+    }
+
+    internal abstract class AbstractBracePairsService : IBracePairsService
+    {
+        private readonly Dictionary<int, int> _bracePairKinds = new();
+
+        protected AbstractBracePairsService(
+            ISyntaxKinds syntaxKinds)
+        {
+            Add(syntaxKinds.OpenBraceToken, syntaxKinds.CloseBraceToken);
+            Add(syntaxKinds.OpenBracketToken, syntaxKinds.CloseBracketToken);
+            Add(syntaxKinds.OpenParenToken, syntaxKinds.CloseParenToken);
+            Add(syntaxKinds.LessThanToken, syntaxKinds.GreaterThanToken);
+
+            return;
+
+            void Add(int? open, int? close)
```

Comment:
Interesting. Is this just because C# and VB would provide different braces to highlight? Why is this done at the abstract instead of implementation?

Reply:
yea, looking at the SyntaxKinds, why not just do this in the constructor of each implementation? It's already <int,int> so each langauge would just add what it needs.